src/main/java/com/structurize/structures/client/TemplateRenderer.java

```diff
@@ -0,0 +1,158 @@
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.gen.structure.template.Template;
+
+import java.util.List;
+
+import static org.lwjgl.opengl.GL11.*;
+
+/**
+ * The renderer for templates.
+ * Holds all information required to render a Template.
+ */
+public class TemplateRenderer
+{
+    private static final float HALF_PERCENT_SHRINK               = 0.995F;
+
+    private final TemplateBlockAccess blockAccess;
+    private final List<TileEntity>    tileEntities;
+    private final List<Entity>        entities;
+    private final TemplateTessellator tessellator;
+    private final BlockPos            primaryBlockOffset;
+
+    /**
+     * Static factory utility method to handle the extraction of the values from the template.
+     *
+     * @param template The template to create an instance for.
+     * @return The renderer.
+     */
+    public static TemplateRenderer buildRendererForTemplate(Template template)
+    {
+        final TemplateBlockAccess blockAccess = new TemplateBlockAccess(template);
+        final List<TileEntity> tileEntities = TemplateUtils.instantiateTileEntities(template, blockAccess);
+        final List<Entity> entities = TemplateUtils.instantiateEntities(template, blockAccess);
+        final TemplateTessellator templateTessellator = new TemplateTessellator();
+        final BlockPos primaryBlockOffset = TemplateUtils.getPrimaryBlockOffset(template);
+
+        return new TemplateRenderer(blockAccess, tileEntities, entities, templateTessellator, primaryBlockOffset);
+    }
+
+    private TemplateRenderer(
+      final TemplateBlockAccess blockAccess,
+      final List<TileEntity> tileEntities,
+      final List<Entity> entities,
+      final TemplateTessellator tessellator,
+      final BlockPos primaryBlockOffset)
+    {
+        this.blockAccess = blockAccess;
+        this.tileEntities = tileEntities;
+        this.entities = entities;
+        this.tessellator = tessellator;
+        this.primaryBlockOffset = primaryBlockOffset;
+
+        this.setup();
+    }
+
+    /**
+     * Sets up the renders VBO
+     */
+    private void setup()
+    {
+        tessellator.startBuilding();
+
+        blockAccess.getTemplate().blocks.stream()
+          .map(b -> TemplateBlockInfoTransformHandler.getInstance().Transform(b))
+          .forEach(b -> Minecraft.getMinecraft().getBlockRendererDispatcher().renderBlock(b.blockState, b.pos, blockAccess, tessellator.getBuilder()));
+
+        tessellator.finishBuilding();
+    }
+
+    /**
+     * Draws an instance of the template at the given position, with the given rotation, and mirroring.
+     *
+     * @param rotation The rotation.
+     * @param mirror The mirroring.
+     * @param drawingOffset The drawing offset.
+     * @param partialTicks The partial tick time.
+     * @param pos The position.
+     */
+    public void draw(final Rotation rotation, final Mirror mirror, final Vector3d drawingOffset, final float partialTicks, final BlockPos pos)
```

Comment:
since the position is done by the drawingOffset, is the final BlockPos pos parameter needed any longer?

Reply:
I am still testing and cleaning up of things that i do not need.
Right now i got most things rendering, but chests and for example the stand of an armor stand are note displaying.